source/common/ssl/ssl_socket.cc

```diff
@@ -263,19 +266,24 @@ std::string SslSocket::getUriSanFromCertificate(X509* cert) {
   return result;
 }
 
-void SslSocket::closeSocket(Network::ConnectionEvent) {
-  if (handshake_complete_ &&
+void SslSocket::halfCloseSocket() {
+  if (!shutdown_sent_ && handshake_complete_ &&
       callbacks_->connection().state() != Network::Connection::State::Closed) {
-    // Attempt to send a shutdown before closing the socket. It's possible this won't go out if
-    // there is no room on the socket. We can extend the state machine to handle this at some point
-    // if needed.
     int rc = SSL_shutdown(ssl_.get());
```

Comment:
hm, I think this cleans up TLS state but I'm not convinced it forwards the FIN on.  I'd think we'd want to wait until SSL_shutdown succeeds  and then send on the FIN.  Alternately we can claim SSL_shutdown is best effort (which svaldez@google tells me is is generally OK) and FIN after.

@PiotrSikora any preference on this one?

Reply:
When you say "then send the FIN" do you mean a TCP-level FIN or a TLS-level FIN (logical FIN)?

Either way, I'm going to add an integration test that has tcp-proxy with TLS on one side and raw TCP on the other and test that half-close gets proxied correctly.

Reply:
TCP FIN :-)

Prior to this PR I hadn't realized TLS logical FIN was a thing, but I know all about proxying TCP FINs :-)

Reply:
In the case of TLS, because it operates at the transport layer, we don't send a TCP FIN at all; boringssl can do that if it wants to.

When the TLS connection is half-closed (at the TLS layer), I'm pretty sure the TCP connection still needs to be bi-directional in case any keys need to be renegotiated or renewed, because the connection could be in this state for an arbitrarily long amount of time.  So calling SSL_shutdown() is all we're supposed to do on a TLS connection to consider it half-closed.

At least that's my understanding of it.  Someone please correct me if I'm wrong.

Reply:
I'd prefer to do the full `close_notify` dance, or at the very least to make sure that we sent `close_notify` before tearing down the connection.

Reply:
According to https://wiki.openssl.org/index.php/Manual:SSL_shutdown(3), SSL_shutdown() sends a close-notify.  But I'm not very familiar with this part of TLS or boring/openssl.  Is there more that should be done to gracefully close the connection?

Reply:
`SSL_shutdown()` is the correct function, but it can fail (`rc == -1`) or might need to be retried (`rc == -1 && (SSL_get_error() ==  SSL_ERROR_WANT_READ || SSL_get_error() ==  SSL_ERROR_WANT_WRITE)`).

`rc == 1` once `close_notify` has been both sent and received.

If we don't want to do that, then we should at least set `SSL_set_quiet_shutdown()` before calling `SSL_shutdown()`, otherwise we might end up in a weird state.